python/tchannel/tornado/peer.py

```diff
@@ -458,34 +506,79 @@ def send(self, arg1, arg2, arg3, traceflag=False, headers=None):
-        response = yield connection.send_request(
-            Request(
-                service=self.service,
-                argstreams=[InMemStream(endpoint), arg2, arg3],
-                id=message_id,
-                headers=headers,
-                tracing=Trace(
-                    name=endpoint,
-                    trace_id=trace_id,
-                    parent_span_id=parent_span_id,
-                    endpoint=Endpoint(self.peer.host,
-                                      self.peer.port,
-                                      self.service),
-                    traceflags=traceflag,
-                )
+        for k, v in self.headers.iteritems():
+            headers.setdefault(k, v)
+
+        peer = self.peer
+        connection = yield peer.connect()
+
+        request = Request(
+            service=self.service,
+            argstreams=[InMemStream(endpoint), arg2, arg3],
+            id=connection.next_message_id(),
+            headers=headers,
+            tracing=Trace(
+                name=endpoint,
+                trace_id=trace_id,
+                parent_span_id=parent_span_id,
+                endpoint=Endpoint(self.peer.host,
+                                  self.peer.port,
+                                  self.service),
+                traceflags=traceflag,
             )
         )
 
+        response = None
+        # black list to record all used peers, so they aren't chosen again.
+        blacklist = set()
+        # only retry on non-stream request
+        if request.is_streaming_request:
+            retry_times = 1
+            timeout_per_req = 0
+
+        # mac number of times to retry 3.
+        for num_of_retry in range(retry_times):
+            try:
+                response = yield self._retry_send(
+                    connection, request, timeout_per_req)
+                break
+            except (ProtocolError, TimeoutError) as e:
+                request.set_exception(e)
+                if not request.should_retry_on_error(e) or self._hostport:
+                    raise
+
+                if num_of_retry != retry_times - 1:
+                    # delay further retry
+                    yield gen.sleep(retry_delay)
+                else:
+                    raise
+
+                blacklist.add(peer.hostport)
+                # find new peer
+                peer = self.peer.tchannel.peers.choose(
+                    hostport=self._hostport,
+                    score_threshold=self._score_threshold,
+                    blacklist=blacklist,
+                )
+                if not peer:
+                    raise
+                connection = yield peer.connect()
+                # roll back request
+                request.rewind(connection.next_message_id())
+
         log.debug("Got response %s", response)
         raise gen.Return(response)
 
+    @gen.coroutine
+    def _retry_send(self, connection, req, timeout_per_req):
```

Comment:
This function's implementation has nothing to do with retries. Call it something else?